// react-flip-numbers/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var reactSimpleAnimate = require('react-simple-animate');

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    keys.push.apply(keys, Object.getOwnPropertySymbols(object));
  }

  if (enumerableOnly) keys = keys.filter(function (sym) {
    return Object.getOwnPropertyDescriptor(object, sym).enumerable;
  });
  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(source, true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(source).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

var commonAnimateStyle = {
  position: 'absolute',
  height: '100%',
  transformStyle: 'preserve-3d'
};
var easeType = 'cubic-bezier(0.19, 1, 0.22, 1)';
var revolutionDegrees = 360;
var resetRouteCounter = 1000;
var numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
var rotateDegreePerNumber = 36;

var calculateDegrees = function calculateDegrees(rotateCounter, activeNumber) {
  var animateDegree = numbers.findIndex(function (v) {
    return v === activeNumber;
  }) * rotateDegreePerNumber;
  var amountDegree = rotateCounter * revolutionDegrees;
  return _objectSpread2({}, activeNumber === 0 ? {
    rotateCounter: rotateCounter > resetRouteCounter ? 0 : rotateCounter + 1
  } : null, {
    degree: amountDegree - animateDegree
  });
};

var FlipNumber =
/*#__PURE__*/
function (_React$Component) {
  _inherits(FlipNumber, _React$Component);

  function FlipNumber() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, FlipNumber);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(FlipNumber)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      degree: 0,
      rotateCounter: 0 // eslint-disable-line react/no-unused-state

    });

    _defineProperty(_assertThisInitialized(_this), "updateNumber", function () {
      _this.setState(function (_ref) {
        var rotateCounter = _ref.rotateCounter;
        return calculateDegrees(rotateCounter, _this.props.activeNumber);
      });
    });

    return _this;
  }

  _createClass(FlipNumber, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      this.updateNumberTimeout = setTimeout(function () {
        return _this2.updateNumber();
      }, 50 * this.props.position);
    }
  }, {
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.activeNumber !== this.props.activeNumber || nextProps.height !== this.props.height || nextProps.width !== this.props.width || this.state.degree === 0 || nextProps.play !== this.props.play;
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.updateNumberTimeout);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          activeNumber = _this$props.activeNumber,
          height = _this$props.height,
          color = _this$props.color,
          background = _this$props.background,
          width = _this$props.width,
          perspective = _this$props.perspective,
          duration = _this$props.duration,
          play = _this$props.play,
          delay = _this$props.delay,
          length = _this$props.length,
          position = _this$props.position,
          _this$props$numberSty = _this$props.numberStyle,
          numberStyle = _this$props$numberSty === void 0 ? {} : _this$props$numberSty;
      var degree = this.state.degree;
      var viewPortSize = {
        width: "".concat(width, "px"),
        height: "".concat(height + 3, "px")
      };
      var halfElementHeight = height / 2;
      var translateZ = halfElementHeight + height;
      return React.createElement("span", {
        style: _objectSpread2({}, viewPortSize, {
          perspective: perspective,
          overflow: 'hidden',
          display: 'inline-block',
          textAlign: 'left',
          height: height
        }),
        "aria-hidden": true
      }, React.createElement(reactSimpleAnimate.Animate, _extends({
        tag: "span",
        play: play,
        start: _objectSpread2({}, commonAnimateStyle),
        end: _objectSpread2({}, commonAnimateStyle, {
          transform: "rotateX(".concat(degree, "deg)")
        })
      }, {
        easeType: easeType,
        duration: duration,
        delay: delay
      }, {
        render: function render(_ref2) {
          var style = _ref2.style;
          return React.createElement("span", {
            style: style
          }, numbers.map(function (n, i) {
            return React.createElement("span", {
              style: _objectSpread2({}, viewPortSize, {
                height: height,
                lineHeight: "".concat(height, "px"),
                fontSize: "".concat(height - 1, "px"),
                position: 'absolute',
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                textAlign: 'center',
                WebkitFontSmoothing: 'antialiased',
                color: color,
                background: background,
                backfaceVisibility: 'hidden',
                WebkitBackfaceVisibility: 'hidden',
                transform: "rotateX(".concat(rotateDegreePerNumber * i, "deg) translateZ(").concat(translateZ, "px)")
              }, numberStyle),
              key: "".concat(rotateDegreePerNumber * i)
            }, n);
          }));
        }
      })), React.createElement("span", {
        data: length - position,
        style: _objectSpread2({}, viewPortSize, {
          height: height,
          lineHeight: "".concat(height, "px"),
          fontSize: "".concat(height - 1, "px"),
          left: "".concat(length - position > 4 ? 0.25 : 0, "px"),
          // hacky fix for weird misalignment in Chrome.
          position: 'absolute',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          textAlign: 'center',
          WebkitFontSmoothing: 'antialiased',
          color: color,
          background: background,
          transform: "rotateX(0deg) translateZ(".concat(translateZ, "px)"),
          visibility: 'hidden'
        }, numberStyle)
      }, activeNumber));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(_ref3, _ref4) {
      var activeNumber = _ref3.activeNumber;
      var rotateCounter = _ref4.rotateCounter;
      return calculateDegrees(rotateCounter, activeNumber);
    }
  }]);

  return FlipNumber;
}(React.Component);

var FlipNumbers =
/*#__PURE__*/
function (_React$Component) {
  _inherits(FlipNumbers, _React$Component);

  function FlipNumbers() {
    _classCallCheck(this, FlipNumbers);

    return _possibleConstructorReturn(this, _getPrototypeOf(FlipNumbers).apply(this, arguments));
  }

  _createClass(FlipNumbers, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.numbers !== this.props.numbers || nextProps.height !== this.props.height || nextProps.width !== this.props.width || nextProps.duration !== this.props.duration || nextProps.delay !== this.props.delay || nextProps.play !== this.props.play;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          numbers = _this$props.numbers,
          nonNumberStyle = _this$props.nonNumberStyle,
          numberStyle = _this$props.numberStyle,
          height = _this$props.height,
          width = _this$props.width,
          color = _this$props.color,
          background = _this$props.background,
          perspective = _this$props.perspective,
          duration = _this$props.duration,
          animate = _this$props.animate,
          play = _this$props.play,
          delay = _this$props.delay;
      var numberCounter = 0;
      return React.createElement("section", {
        style: {
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center'
        },
        "aria-label": numbers
      }, Array.from(numbers).map(function (n, key) {
        var nonNumber = React.createElement("span", {
          "aria-hidden": true,
          style: nonNumberStyle,
          key: numberCounter
        }, n);

        if (animate) {
          numberCounter += 1;
          return !Number.isNaN(parseInt(n, 10)) ? React.createElement(FlipNumber, _extends({
            key: key,
            height: height,
            width: width,
            color: color,
            background: background,
            perspective: perspective,
            duration: duration,
            play: play,
            delay: delay,
            numberStyle: numberStyle
          }, {
            position: numberCounter,
            length: numbers.length,
            activeNumber: parseInt(n, 10)
          })) : nonNumber;
        }

        return !Number.isNaN(parseInt(n, 10)) ? React.createElement("span", {
          "aria-hidden": true,
          style: {
            padding: 0
          },
          className: nonNumberStyle,
          key: numberCounter
        }, n) : nonNumber;
      }));
    }
  }]);

  return FlipNumbers;
}(React.Component);

_defineProperty(FlipNumbers, "defaultProps", {
  perspective: 500,
  duration: 0.3,
  animate: true,
  play: false,
  delay: 0,
  nonNumberStyle: {},
  numberStyle: {}
});

exports.FlipNumber = FlipNumber;
exports.default = FlipNumbers;
